### PR TITLE
Parallel directory iterator initialization

### PR DESCRIPTION
**Pull request by GPhilo**
Sped up DirectoryIterator's __init__ execution by parallelizing via multiprocessing the samples counting and extraction of samples labels and paths.
This can speed up execution significantly when dealing with big datasets.

### CODE REVIEW

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -821,6 +823,71 @@ def next(self):
...
+            for extension in white_list_formats:
+                if fname.lower().endswith('.' + extension):
+                    is_valid = True
+                    break
+            if is_valid:
+                samples += 1
+    return samples
+
+
+def _list_valid_filenames_in_directory(subdir, directory, white_list_formats,
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Why not just pass the target directory? This is all your need to perform this task.

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -821,6 +823,71 @@ def next(self):
         return batch_x, batch_y
 
 
+def _count_valid_files_in_directory(dirpath, white_list_formats, follow_links):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Use `directory` for consistency

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -580,8 +582,8 @@ def random_transform(self, x):
 
         if shear != 0:
             shear_matrix = np.array([[1, -np.sin(shear), 0],
-                                    [0, np.cos(shear), 0],
-                                    [0, 0, 1]])
+                                     [0, np.cos(shear), 0],
+                                     [0, 0, 1]])
```
Comment (fchollet) [inline keras/preprocessing/image.py]: The indent is meant to match to upper logical line, not the semantics of np arrays

Comment (GPhilo) [inline keras/preprocessing/image.py]: Sorry, I think this was an accidental automatic correction from autopep8

File: keras/preprocessing/image.py (hunk 1 above)
Comment (GPhilo) [inline keras/preprocessing/image.py]: Which function are you referring to?
If it's _list_valid_[...] then I passed both subdir and directory to make it more explicit that the filenames in the returned list are relative to directory.
My original version took just the absolute path to subdir, but I think this is clearer and simpler to document (although possibly prone to errors if subdir is not actually contained in directory).
What do you think?

Comment (fchollet) [inline keras/preprocessing/image.py]: I would use the absolute path to the directory where you are listing files, for simplicity.

**Comment by GPhilo** [comment]
I'm thinking of whether this is better done using threads instead of processes.
At the moment processes are hardcoded, so that the interface of DirectoryIterator does not change, but I could add a flag to `__init__` that could specify whether to use threads or processes.
The downside is then though that `ImageDataGenerator::flow_from_directory` should also have the same flag.

**Comment by fchollet** [comment]
> I'm thinking of whether this is better done using threads instead of processes.

You are right that we should do this with threads, not processes. Threads are best when the bottleneck is I/O (like in this case). Processes are best when the bottleneck is compute (not in this case), and are more likely to cause technical issues (leaks, etc).

**Comment by fchollet** [comment]
Can you make sure all the new code is tested (see `test_directory_iterator` unit test)?

**Comment by fchollet** [comment]
LGTM

**Comment by oak-tree** [comment]
Hey! this is great!
@GPhilo 
Small remark - or more  accurate, "leftover cleaning". I think the `_recursive_list` function which is define at [DirectoryIterator \_\_init\_\_ method](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L982) is no longer been use since  the job is done now by both `_list_valid_filenames_in_directory` and `_count_valid_files_in_directory`

**Comment by GPhilo** [comment]
You're absolutely right. I actually realized this shortly after the PR got merged.
At the moment I'm a bit caught up with things, so if you want you could open a PR with the fix. Otherwise I'll probably do it myself during the weekend.

**Pull request merged**